### PR TITLE
Add PyTorch DPT decoder module with shared head and task-specific decoders

### PR DESCRIPTION
**Pull request by bingyic**
## Summary

Add `pytorch/decoders.py` implementing the DPT dense prediction decoder hierarchy in PyTorch, mirroring the [Scenic/Flax decoder structure](https://source.corp.google.com/piper///depot/google3/third_party/py/scenic/projects/dense_features/models/decoders.py).

## Architecture

The module follows a shared-head pattern where a common DPT backbone is subclassed for each task:

- **`DPTHead`**: Shared backbone performing feature reassembly at multiple scales (ReassembleBlocks) and bottom-up fusion (FeatureFusionBlocks).
- **`Decoder`**: Base class wrapping a DPTHead with a common interface.
- **`SegmentationDecoder(Decoder)`**: Per-pixel class logits (`num_classes` output channels).
- **`DepthDecoder(Decoder)`**: Monocular depth prediction (1-channel output).
- **`NormalsDecoder(Decoder)`**: Surface normals prediction (3-channel output).

## Weight Loading

`load_decoder_weights()` loads pre-converted `.npz` checkpoints (produced by `scripts/convert_segmentation_checkpoint.py`) with backward-compatible key remapping from the legacy flat format.

## Usage

```python
from tips.pytorch.decoders import SegmentationDecoder, load_decoder_weights

decoder = SegmentationDecoder(num_classes=150, input_embed_dim=1024)
load_decoder_weights(decoder, 'path/to/checkpoint.npz')
logits = decoder(intermediate_features, image_size=(480, 640))
```

### CODE REVIEW

File: pytorch/decoders.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    model: Decoder,
+    checkpoint_path: str,
+) -> Decoder:
+  """Load pre-converted PyTorch weights into a Decoder.
+
+  The checkpoint should be an .npz file where keys match the model's
+  state_dict parameter names and values are NumPy arrays in PyTorch layout
+  (already transposed from Flax/JAX format).
+
+  Use ``scripts/convert_segmentation_checkpoint.py`` to produce these
```
Comment (koertchen) [inline pytorch/decoders.py]: please remove this internal detail

File: pytorch/decoders.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+    # 3. Upsample to target resolution
+    if image_size is not None:
+      x = F.interpolate(
+          x, size=image_size, mode="bilinear", align_corners=False
+      )
+    return x
+
+
+# ---------------------------------------------------------------------------
+# Task-specific decoders (Refactored as Thin Wrappers)
```
Comment (koertchen) [inline pytorch/decoders.py:256]: nit. remove "(Refactored as Thin Wrappers)"

File: pytorch/decoders.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""DPT decoder heads for dense prediction tasks.
+
+This module provides a shared DPT backbone (ReassembleBlocks + fusion) and
+task-specific decoder subclasses for segmentation, depth, and surface normals,
+mirroring the Scenic/Flax decoder hierarchy.
```
Comment (koertchen) [inline pytorch/decoders.py]: Remove mention of our scenic/flax internals.

File: pytorch/decoders.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+        residual = F.interpolate(
+            residual,
+            size=x.shape[2:],
+            mode="bilinear",
+            align_corners=False,
+        )
+      residual = self.residual_unit(residual)
+      x = x + residual
+    x = self.main_unit(x)
+    # Upsample 2x with align_corners=True (matches Scenic reference).
```
Comment (koertchen) [inline pytorch/decoders.py]: also here

File: pytorch/decoders.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+# ---------------------------------------------------------------------------
+
+
+class SegmentationDecoder(Decoder):
+  """Decoder for semantic segmentation."""
+  def __init__(self, num_classes: int = 150, **kwargs) -> None:
+    super().__init__(out_channels=num_classes, **kwargs)
+
+
+class DepthDecoder(Decoder):
```
Comment (koertchen) [inline pytorch/decoders.py:266]: DepthDecoder actually is classification based, not regression:

class DepthDecoder(Decoder):
  """Decoder for monocular depth prediction using classification bins."""
  def __init__(
      self, min_depth: float = 0.001, max_depth: float = 10.0, **kwargs
  ) -> None:
    super().__init__(out_channels=256, **kwargs)
    self.min_depth = min_depth
    self.max_depth = max_depth
    self.register_buffer(
        "bin_centers", torch.linspace(min_depth, max_depth, 256)
    )
  def forward(
      self,
      intermediate_features: List[Tuple[torch.Tensor, torch.Tensor]],
      image_size: Optional[Tuple[int, int]] = None,
  ) -> torch.Tensor:
    logits = super().forward(intermediate_features, image_size=image_size)
    # Apply ReLU and shift
    logits = torch.relu(logits) + self.min_depth
    # Normalize to probabilities along the channel dimension
    probs = logits / torch.sum(logits, dim=1, keepdim=True)
    # Compute expectation: sum(prob * bin_center)
    depth_map = torch.einsum(
        "bchw,c->bhw", probs, self.bin_centers.to(logits.device)
    )
    return depth_map.unsqueeze(1)
